homeassistant/helpers/llm.py

```diff
@@ -0,0 +1,408 @@
+    device_id: str | None = None,
+    assistant: str | None = None,
+) -> str:
+    """Call a LLM tool, parse args and return the response."""
+    if (tools := hass.data.get(DATA_KEY)) is None:
+        tools = {}
+
+    if context is None:
+        context = Context()
+
+    try:
+        tool = tools[tool_name]
+        tool_args = json.loads(json_args)
+        response = await tool.async_call(
+            hass,
+            platform,
+            text_input,
+            context,
+            language,
+            agent_id,
+            conversation_id,
+            device_id,
+            assistant,
+            **tool_args,
+        )
+        json_response = json.dumps(response)
+
+    except Exception as e:  # pylint: disable=broad-exception-caught
+        response = {"error": type(e).__name__}
+        if str(e):
+            response["error_text"] = str(e)
+        json_response = json.dumps(response)
+
+    return json_response
+
+
+EXPORTED_ATTRIBUTES = [
+    "device_class",
+    "message",
+    "all_day",
+    "start_time",
+    "end_time",
+    "location",
+    "description",
+    "hvac_modes",
+    "min_temp",
+    "max_temp",
+    "fan_modes",
+    "preset_modes",
+    "swing_modes",
+    "current_temperature",
+    "temperature",
+    "target_temp_high",
+    "target_temp_low",
+    "fan_mode",
+    "preset_mode",
+    "swing_mode",
+    "hvac_action",
+    "aux_heat",
+    "current_position",
+    "current_tilt_position",
+    "latitude",
+    "longitude",
+    "percentage",
+    "direction",
+    "oscillating",
+    "available_modes",
+    "max_humidity",
+    "min_humidity",
+    "action",
+    "current_humidity",
+    "humidity",
+    "mode",
+    "faces",
+    "total_faces",
+    "min",
+    "max",
+    "step",
+    "min_color_temp_kelvin",
+    "max_color_temp_kelvin",
+    "min_mireds",
+    "max_mireds",
+    "effect_list",
+    "supported_color_modes",
+    "color_mode",
+    "brightness",
+    "color_temp_kelvin",
+    "color_temp",
+    "hs_color",
+    "rgb_color",
+    "xy_color",
+    "rgbw_color",
+    "rgbww_color",
+    "effect",
+    "sound_mode_list",
+    "volume_level",
+    "is_volume_muted",
+    "media_content_type",
+    "media_duration",
+    "media_position",
+    "media_title",
+    "media_artist",
+    "media_album_name",
+    "media_track",
+    "media_series_title",
+    "media_season",
+    "media_episode",
+    "app_name",
+    "sound_mode",
+    "shuffle",
+    "repeat",
+    "source",
+    "options",
+    "battery_level",
+    "available_tones",
+    "elevation",
+    "rising",
+    "fan_speed_list",
+    "fan_speed",
+    "status",
+    "cleaned_area",
+    "operation_list",
+    "operation_mode",
+    "away_mode",
+    "temperature_unit",
+    "pressure",
+    "pressure_unit",
+    "wind_speed",
+    "wind_speed_unit",
+    "dew_point",
+    "cloud_coverage",
+    "persons",
+]
+
+
+def _format_state(hass: HomeAssistant, entity_state: State) -> dict[str, Any]:
```

Comment:
I think we should tackle returning data in a follow up PR. I think the API needs some additional consideration. In the [architecture discussion](https://github.com/home-assistant/architecture/discussions/1068) the proposal was to tackle querying entities in follow up work. I agree that giving the LLM useful data about attributes is a good idea, though i am not sure yet that this should be a set of attributes defined in this file.

Reply:
It is not only for querying. Some intents return the list of states it affected and the list of states id could not.
Returning `state.as_dict()` is too expensive in terms of input tokens, `state.name()` is too little information, so this was a compromise I'd come up with. I also don't like the hardcoded list of 'meaningful' attributes.
I will remove them completely for now, and I agree, let's leave it for a follow-up PR.